### PR TITLE
[#336] As a developer, I can upload the build to App Store Connect with Fastlane Swift

### PR DESCRIPTION
**Pull request by vnntsu**
#336

## What happened

- Create `uploadToAppStore` function at `Distribution`
- Support making App Store API Key:
  - Add `appstoreConnectAPIKey` in `Secret`
  - Add `appStoreKeyId` and `appStoreKeyId` to `Constant`

## Insight

- Rearrange `Constant` `MARK` by alphabet order
- Create lane `buildAndUploadToAppStoreLane`
 
## Proof Of Work

Screen Shot 2022-09-27 at 16 51 08

Screen Shot 2022-09-27 at 16 58 50

### CODE REVIEW

File: fastlane/Helpers/Distribution.swift [hunk 1]
```diff
@@ -25,4 +25,27 @@ enum Distribution {
...
     }
+
+    static func uploadToAppStore() {
+        let environment: Constant.Environment = .production
+        let ipaPath = makeIpaPath(environment: environment)
+        appStoreConnectApiKey(
+            keyId: Constant.appStoreKeyId,
+            issuerId: Constant.appStoreIssuerId,
+            keyContent: .userDefined(Secret.appstoreConnectAPIKey),
+            isKeyContentBase64: .userDefined(true) // Check if the AppStore Connect API Key is base64
```
Comment (blyscuit) [inline fastlane/Helpers/Distribution.swift:36]:
```suggestion
            isKeyContentBase64: .userDefined(true) // Use AppStore Connect API Key in base64 encoded
```

**Review by blyscuit — APPROVED** [review]

File: fastlane/Helpers/Distribution.swift (hunk 1 above)
Comment (vnntsu) [reply fastlane/Helpers/Distribution.swift:36]: @blyscuit We have two ways to procedure API key content, so I think it's better if the user checks to ensure they're using base64 or not. We don't need to force using base64 format, IMO.

File: fastlane/Helpers/Distribution.swift [hunk 2]
```diff
@@ -25,4 +25,27 @@ enum Distribution {
             debug: .userDefined(true)
         )
     }
+
+    static func uploadToAppStore() {
+        let environment: Constant.Environment = .production
+        let ipaPath = makeIpaPath(environment: environment)
+        appStoreConnectApiKey(
```
Comment (suho) [inline fastlane/Helpers/Distribution.swift:32]: Let's separate this into a new function. I might use this for downloading the dSYM file. How about create

```swift
enum AppStoreAuthentication { static func connectAPIKey() }
```

File: fastlane/Helpers/Distribution.swift [hunk 3]
```diff
@@ -25,4 +25,27 @@ enum Distribution {
...
+            appIdentifier: .userDefined(environment.bundleId),
+            ipa: .userDefined(ipaPath),
+            skipScreenshots: .userDefined(true),
+            skipMetadata: .userDefined(true),
+            force: .userDefined(true),
+            runPrecheckBeforeSubmit: .userDefined(false)
+        )
+    }
+
+    static private func makeIpaPath(environment: Constant.Environment) -> String {
```
Comment (suho) [inline fastlane/Helpers/Distribution.swift:48]: I would prefer

```swift
private static func makeIPAPath(...
```

File: fastlane/Helpers/Distribution.swift [hunk 4]
```diff
@@ -25,4 +25,27 @@ enum Distribution {
             debug: .userDefined(true)
         )
     }
+
+    static func uploadToAppStore() {
+        let environment: Constant.Environment = .production
```
Comment (suho) [inline fastlane/Helpers/Distribution.swift:30]: How about this?
```swift
static func uploadToAppStore(environment: Constant.Environment = .production)
```

We can support upload staging build to AppStore as well, some project did this.

**Review by ducbm051291 — APPROVED** [review]
I think Su.V's suggestion is reasonable. The rest LGTM

File: fastlane/Constants/Constant.swift [hunk 5]
```diff
@@ -8,30 +8,37 @@
...
+    // MARK: - App Store
+
+    static let appStoreKeyId: String = "<#appStoreKeyId#>"
+    static let appStoreIssuerId: String = "<#appStoreIssuerId#>"
+
+    // MARK: - Firebase
+
+    static let stagingFirebaseAppId: String = "<#stagingFirebaseAppId#>"
+    static let productionFirebaseAppId: String = "<#productionFirebaseAppId#>"
+    static let firebaseTesterGroups: String = "<#group1#>, <#group2#>"
```
Comment (minhnimble) [inline fastlane/Constants/Constant.swift:20]: Just curious, why don't we turn this into an array of Strings instead? 🤔
```suggestion
    static let firebaseTesterGroups: [String] = ["<#group1#>", "<#group2#>"]
```

Comment (ducbm051291) [reply fastlane/Constants/Constant.swift:20]: I think it's because in the final, it must convert to a string for Fastlane to send

Comment (vnntsu) [reply fastlane/Constants/Constant.swift:20]: @minhnimble Good concern, you can check the [doc](https://firebase.google.com/docs/app-distribution/ios/distribute-fastlane) of firebase distribution. It is a string, so for me, I don't want extra logic to join them when passing to the function. What do you think?

Screen Shot 2022-09-28 at 16 08 34

File: fastlane/Constants/Secret.swift [hunk 6]
```diff
@@ -12,4 +12,6 @@ enum Secret {
     static let keychainPassword: String = "<#keychainPassword#>"
 
     static let firebaseCLIToken: String = "<#firebaseCLIToken#>"
+
+    static let appstoreConnectAPIKey = "<#appstoreConnectAPIKey#>"
```
Comment (blyscuit) [inline fastlane/Constants/Secret.swift:16]: This is the only constant without `: String`. I think we don't need the `: String` at all for these cases?

Comment (vnntsu) [reply fastlane/Constants/Secret.swift:16]: @blyscuit Updated as this [thread](https://nimble-co.slack.com/archives/C418DBL2W/p1664364433445829) too in [f419874](https://github.com/nimblehq/ios-templates/pull/359/commits/f419874843448853c24199fb176948596b703a8b) as

**Review by suho — APPROVED** [review]
Approved

File: fastlane/Constants/Constant.swift (hunk 5 above)
Comment (minhnimble) [reply fastlane/Constants/Constant.swift]: noted, thanks for the sharing @vnntsu 👍

**Pull request merged**